### ISSUE TITLE
Allow the option to archive with a headless browser

### ISSUE THREAD

**Issue by hellodword** [go-shiori/obelisk#14]
Just like archivebox, I think archivebox is very nice, but there're two issues:

1. slow, not a big deal;
2. custom automation for special pages (lazy loading for example), [this issue](https://github.com/ArchiveBox/ArchiveBox/issues/51) is working on it.

And I found a great golang lib [rod](https://github.com/go-rod/rod), how about adding a mode of using headless (or headful, it depends) chromium?

**Comment by waybackarchiver** [comment]
That is a fantastic idea. Given the original requirement, we implemented similar features in [screenshot](https://github.com/wabarc/screenshot), but it is still not what you expected.

Perhaps we can take things further and develop a piecemeal approach here.

**Comment by hellodword** [comment]
The biggest challenge for me is developing or choosing a script and its interpreter.

I have no experience of this before, but rod has a good [api](https://github.com/go-rod/rod/blob/14ebb72947cc99e916ed97897473be2b6c64f39d/page_eval.go#L195-L197). :smile: 

I will try to implement this, I really prefer this mode, dealing with all elements is too hard.

**Comment by fmartingr** [comment]
If this happens, I'd like the feature to be optional if it require a lot /complex of external dependencies. I'm trying to maintain shiori as simple as possible and we just get rid of CGO switching the SQLite driver, and since wasm is obsolete and I have to start replacing it with Obelisk, the more seamless the experience it can be, the better :)

**Comment by hellodword** [comment]
> If this happens, I'd like the feature to be optional if it require a lot /complex of external dependencies. I'm trying to maintain shiori as simple as possible and we just get rid of CGO switching the SQLite driver, and since wasm is obsolete and I have to start replacing it with Obelisk, the more seamless the experience it can be, the better :)

I'm also a big fan of CGo-free and fewer dependencies, chromedp and rod are based on  [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/), without CGo or tons of dependencies. :smiley:

**Comment by waybackarchiver** [comment]
@fmartingr Please don't be worried about complex external dependencies. Perhaps we can look forward to the given works.

Anyway, pr is wecome.

**Comment by hellodword** [comment]
Hey, I created a simple demo.

https://github.com/hellodword/web-archiving-with-headless-chromium-demo

```sh
env rod=show,bin=/path/to/chrome go run .
```

It's very simple, but provides custom `post.js` for hooking and `pre.js` for scroll/click/...

And use singlefile for saving.

**Comment by waybackarchiver** [comment]
This implementation is creative and far superior to our current archiving solution. However, it is heavily dependant on SingleFile, and it appears that obelisk is no longer required. If we stick to this plan, a new project might be a better alternative.

@fmartingr What do you think?

**Comment by hellodword** [comment]
> it is heavily dependant on SingleFile

Right, and it's buggy in this demo. 😂

But it's optional, just like the archivebox, archivebox has multi saving modes, singlefile is only one of it.

The thing I want to show is ability of custom script, and, a highly recommend cdp library of golang, I think it's much better than chromedp.

**Comment by waybackarchiver** [comment]
Appreciate the time and effort. Personally, I prefer the option of trying to inject the script in headless over the one implemented in the [screenshot](https://github.com/wabarc/screenshot/blob/77c4be05f0fc5739102815bcbcec8f30d8be9a97/screenshot.go#L280-L296) project.

It appears that making it an option would be reasonable, so if SingleFile is added as a browser extension, I would prefer to put the gitmodule in the `.github/thridparty` directory.

An example of archiving results using `screenshot`:

![image](https://user-images.githubusercontent.com/66856220/155757669-9dd73807-40f2-4335-bbac-512b6b450746.png)

**Comment by hellodword** [comment]
Right, it was a demo so I directly use singlefile as an embedded dependency, it could or should be act as a `plugins`.

I think nowadays archiving tool do not necessarily need a chromium, but need ability of scripting extension, one reason is there're too much anti-bot stuff (captcha, WAF, and so on) on the internet.

**Comment by Katarn** [comment]
It seems to me that the ideal solution would be the ability to prepare the page for saving not on the server, but on the client. And send it to the server.

Now a lot of sites use dynamic image loading, captcha checking, they load comments only if you scroll the page to them (and comments are sometimes more interesting than the article itself), they don’t load all comments (hide discussion threads until you force them to open). Lots of dynamics. Therefore, it is better to save the page after having previously examined it with your own eyes, that all that is needed is loaded and displayed. There is no universal solution here, so it is preferable to inspect the page yourself.

I just looked into my Pocket archive and it became very sad - many domains are already partitioned, there are no sites. And the pages themselves (at a premium tariff) are far from being completely saved, sometimes they don’t even have text. And now I'm looking for a solution to this problem. I have now started saving pages through SingleFile, but if you tie it to shiori, it will be just the perfect bookmark manager.

At the same time, I would like shiori not to save the text to its database (perhaps only for a quick search), but always retrieve it again from the saved page. Because text content recognition algorithms will always improve, and the content stored in the database may be incorrectly recognized and no longer relevant from the new version of the application.

**Comment by waybackarchiver** [comment]
@Katarn Thank you for your offer, it's a fantastic idea. As intended, `obelisk` should support both `headless` and `non-headless` mode for archiving webpage. 

> if you tie it to shiori, it will be just the perfect bookmark manager.

Makes `shiori` work with `obelisk` is related to https://github.com/go-shiori/shiori/issues/353
